cirq/qis/measures.py

```diff
@@ -38,36 +56,211 @@ def fidelity(state1: np.ndarray, state2: np.ndarray) -> float:
+
+
+def _numpy_arrays_to_state_vectors_or_density_matrices(
+    state1: np.ndarray,
+    state2: np.ndarray,
+    qid_shape: Optional[Tuple[int, ...]],
+    validate: bool,
+    atol: float,
+) -> Tuple[np.ndarray, np.ndarray]:
+    if state1.ndim > 2 or state1.ndim == 2 and state1.shape[0] != state1.shape[1]:
+        # State tensor, convert to state vector
+        state1 = np.reshape(state1, (np.prod(state1.shape),))
+    if state2.ndim > 2 or state2.ndim == 2 and state2.shape[0] != state2.shape[1]:
+        # State tensor, convert to state vector
+        state2 = np.reshape(state2, (np.prod(state2.shape),))
+    if state1.ndim == 2 and state2.ndim == 2:
+        # Must be square matrices
+        if state1.shape == state2.shape:
+            if qid_shape is None:
+                # Ambiguous whether state tensor or density matrix
+                raise ValueError(
+                    'The qid shape of the given states is ambiguous.'
+                    'Try specifying the qid shape explicitly or '
+                    'using a wrapper function like cirq.density_matrix.'
+                )
+            if state1.shape == qid_shape:
+                # State tensors, convert to state vectors
+                state1 = np.reshape(state1, (np.prod(qid_shape),))
+                state2 = np.reshape(state2, (np.prod(qid_shape),))
+        elif state1.shape[0] < state2.shape[0]:
+            # state1 is state tensor and state2 is density matrix.
+            # Convert state1 to state vector
+            state1 = np.reshape(state1, (np.prod(state1.shape),))
+        else:  # state1.shape[0] > state2.shape[0]
+            # state2 is state tensor and state1 is density matrix.
+            # Convert state2 to state vector
+            state2 = np.reshape(state2, (np.prod(state2.shape),))
+    elif state1.ndim == 2 and state2.ndim < 2 and np.prod(state1.shape) == np.prod(state2.shape):
+        # state1 is state tensor, convert to state vector
+        state1 = np.reshape(state1, (np.prod(state1.shape),))
+    elif state1.ndim < 2 and state2.ndim == 2 and np.prod(state1.shape) == np.prod(state2.shape):
+        # state2 is state tensor, convert to state vector
+        state2 = np.reshape(state2, (np.prod(state2.shape),))
+
+    if validate:
+        dim1 = state1.shape[0] if state1.ndim == 2 else np.prod(state1.shape)
+        dim2 = state2.shape[0] if state2.ndim == 2 else np.prod(state2.shape)
+        if dim1 != dim2:
+            raise ValueError('Mismatched dimensions in given states: ' f'{dim1} and {dim2}.')
+        if qid_shape is None:
+            qid_shape = (dim1,)
+        else:
+            expected_dim = np.prod(qid_shape)
+            if dim1 != expected_dim:
+                raise ValueError(
+                    'Invalid state dimension for given qid shape: '
+                    f'Expected dimension {expected_dim} but '
+                    f'got dimension {dim1}.'
+                )
+        if state1.ndim == 2:
```

Comment:
I'd combine these two ifs into a single for to make it more concise

Reply:
Done.